lib/puppet/type/file.rb

```diff
@@ -434,7 +436,8 @@ def bucket
       fail _("Can not find filebucket for backups without a catalog")
     end
 
-    unless catalog and filebucket = catalog.resource(:filebucket, backup) or backup == "puppet"
+    filebucket = catalog.resource(:filebucket, backup) if catalog
+    unless catalog && filebucket || backup == "puppet"
```

Comment:
I find `unless` confusing with multiple `&&` and `||` and taking operator precedence into account. To be honest I'm not sure of the precedence there. I find it easier to follow something like (not sure I have the parentheses correct)

```ruby
if !catalog || (!filebucket && bucket != 'puppet')
  fail ...
```